core/src/sumeragi/mod.rs

```diff
@@ -54,13 +59,13 @@ struct LastUpdateMetricsData {
 #[derive(Clone)]
 pub struct SumeragiHandle {
     public_wsv: Arc<Mutex<WorldStateView>>,
-    message_sender: mpsc::SyncSender<MessagePacket>,
     metrics: Metrics,
     last_update_metrics_mutex: Arc<Mutex<LastUpdateMetricsData>>,
-    network: IrohaNetwork,
+    get_connected_peers_procedure: SumeragiGetCurrentlyConnectedProcedure,
     kura: Arc<Kura>,
     queue: Arc<Queue>,
     _thread_handle: Arc<ThreadHandler>,
+    message_sender: mpsc::SyncSender<MessagePacket>, // This must be kept alive until the thread exits, ergo it's after _thread_handle.
```

Comment:
Actually, the better way to express this is by `impl Drop`. Just because `Arc::drop` was called doesn't mean the thread is stopped, just that the refcount got decremented.

Reply:
In fact instead of silently shutting off the `Sender`, it's a good idea to block until the refcount reaches zero and panic if it doesn't in a reasonable amount of time.

Reply:
I disagree. There are no copies of the ThreadHandle arc outside of the SumeragiHandle. So there is a one to one relationship between the number of SumeragiHandle's, Arc<ThreadHandler> and SyncSenders.

Reply:
In which case it's not supposed to be an Arc, which is an owning reference.

Reply:
I kind of agree with you. But it's not for this PR. I have opened an issue. https://github.com/hyperledger/iroha/issues/3584

Reply:
Splitting off into another issue is always good. 

For future reference, it's typically better to use "Reference in new issue", that way you don't go through the template, and refer to the conversation in-context, without having to add things manually.